Frameworks/CoreText/CTLine.mm

```diff
@@ -228,15 +229,19 @@ void CTLineDraw(CTLineRef lineRef, CGContextRef ctx) {
     _CGContextPushBeginDraw(ctx);
     auto popEnd = wil::ScopeExit([ctx]() { _CGContextPopEndDraw(ctx); });
 
+    std::vector<GlyphRunData> runs;
+    CGPoint relativePosition = CGPointZero;
     for (size_t i = 0; i < [line->_runs count]; ++i) {
         _CTRun* curRun = [line->_runs objectAtIndex:i];
         if (i > 0) {
             // Adjusts x position relative to the last run drawn
-            CGPoint curTextPos = CGContextGetTextPosition(ctx);
-            CGContextSetTextPosition(ctx, curTextPos.x + curRun->_relativeXOffset, curTextPos.y);
+            relativePosition.x += curRun->_relativeXOffset;
         }
+        runs.emplace_back(GlyphRunData{ &curRun->_dwriteGlyphRun, relativePosition, (CFDictionaryRef)curRun->_attributes.get() });
+    }
 
-        CTRunDraw(static_cast<CTRunRef>(curRun), ctx, CFRange{});
+    if (!runs.empty()) {
+        _CGContextDrawGlyphRuns(ctx, runs.data(), runs.size());
```

Comment:
There's no point in doing a push begin and pop end if `runs.empty()`. Since we're not managing gstates, it would also be reasonable to get rid of the scopeguard.